### ISSUE TITLE
Question about playback/record with plot synchronization

### ISSUE THREAD

**Issue by larsacus** [audiokit/AudioKit#545]
_tl;dr; I have a pretty simple record/playback use-case that I am using AudioKit for on iOS and am suffering from visual audio plot desynchronization between the analog input/output and the visualization._

## Background/Setup
I am simply recording audio from a device's microphone and storing it in a file (via AVAudioRecorder) while playing the realtime buffered waveform through AudioKit's EZAudioPlot and AKNodeOutputPlot. For playback, I am playing back a recorded m4a from a file using AKAudioPlayer with the same visualization of whatever is being output from the device's speaker for visual feedback of the audio.

Capture: `AKMicrophone` -> `AKFrequencyTracker` -> `AKBooster` -> `AKNodeOutputPlot` -> `EZAudioPlot`
Playback: `AKAudioPlayer` -> `AKFrequencyTracker` -> `AKBooster` -> `AKNodeOutputPlot` -> `EZAudioPlot` 

## Issue
My issue comes when I pause audio (during playback) or interrupt the current audio session  category (during both playback and capture). Namely, when switching the output from the external speaker to the earpiece receiver. When the audio session category has been successfully changed, the visual representation of the audio plot is significantly desynchronized (delayed) from what is physically being played back to the user (or for capture, what is input to the microphone). I've tried completely tearing down the AudioKit stack used for visualization and rebuilding it for every session category change (less than stable), as well as attempting to manually pause the portion of the AudioKit stack that is dealing with the visualization with no success. I cannot get the audio to synchronize back up with the visuals after an audio session category change.

I'm admittedly new to handling audio in any capacity outside of basic sound file playback for button sounds, so I might be a little out of my league. Basically looking for any suggestions on what I'm doing wrong or where to go from here. Let me know if this is the wrong place for this kind of thing as there aren't many issues on this github repo.

Thank you in advance. Cheers! 🍻

**Comment by aure** [comment]
The latency or synchronization with plots seems to be very much affected by the device you're running on.  I don't really get why its not consistent.  Perhaps you've got a clue with the fact that it is made reproducible by the changing of the audio device.  We can leave this as an issue to remind us to be mindful of it when we're looking at the code, but I don't really know what steps I'd take to eradicate this problem.  It's going to take some deep diving.

**Comment by larsacus** [comment]
That's interesting -- I'll see if the synchronization gets better with a faster device (currently testing on an iPhone 5, iOS 9.3.5). It's comforting at least to know my current AudioKit configuration is at least somewhat correct. I just think it's interesting that the device can fully keep up with visualizing the audio as long as the audio category isn't switched, but as soon as it is, the becomes offset. I'm not sure if it helps, but this is with a short (<25s) looping playback file.

🍻

**Comment by aure** [comment]
Any update?

**Comment by larsacus** [comment]
This synchronization doesn't appear to be dependent on the speed of a device. It exhibits the same behavior whether on my iPhone 5 or on my iPhone 6.

**Comment by aure** [comment]
This is working well on my tests.  Can you share your project so I can see what is going on?  Otherwise I don't think much else can be done to close the issue.

**Comment by larsacus** [comment]
One more data point to consider that I might not have emphasized in my original post is that the audio category switch is done simultaneously with an earpiece proximity sensor activation (screen turns off). When the proximity sensor is in the 'off' position (screen on) after proximity has been active, this desynchronization exhibits itself.

**Comment by aure** [comment]
oh wow, yeah, I am not sure I'd know how to debug that without building a project to debug with, which you've kind of already done.  Any chance of sharing, even a skeleton of your current project that shows of the issue would help?

**Comment by larsacus** [comment]
I apologize for the delay on getting you a sample. I haven't forgotten about this and it is still in my backlog. 🍻

**Comment by aure** [comment]
I'm going to close this for now, but feel to reopen when you have the sample ready.
